crates/ruff/src/rules/ruff/typing.rs

```diff
@@ -52,6 +52,15 @@ fn is_known_type(call_path: &CallPath, minor_version: u32) -> bool {
     }
 }
 
+/// Returns a vector of all the expressions in a slice. If the slice is not a
+/// tuple, it will be wrapped in a vector.
+fn resolve_slice_value(slice: &Expr) -> Vec<&Expr> {
+    match slice {
+        Expr::Tuple(ast::ExprTuple { elts: elements, .. }) => elements.iter().collect(),
+        _ => vec![slice],
+    }
+}
```

Comment:
I think you can avoid these allocations by just storing the `slice` `Expr` on the enums directly, and then doing this match later, when you actually need to iterate over the members.

Reply:
Here's a patch of what I'm imagining:

```diff
diff --git a/crates/ruff/src/rules/ruff/typing.rs b/crates/ruff/src/rules/ruff/typing.rs
index f96ea22d8..3ad7d9c9f 100644
--- a/crates/ruff/src/rules/ruff/typing.rs
+++ b/crates/ruff/src/rules/ruff/typing.rs
@@ -1,3 +1,4 @@
+use itertools::Either::{Left, Right};
 use rustpython_parser::ast::{self, Constant, Expr, Operator};
 
 use ruff_python_ast::call_path::CallPath;
@@ -54,10 +55,10 @@ fn is_known_type(call_path: &CallPath, minor_version: u32) -> bool {
 
 /// Returns a vector of all the expressions in a slice. If the slice is not a
 /// tuple, it will be wrapped in a vector.
-fn resolve_slice_value(slice: &Expr) -> Vec<&Expr> {
+fn resolve_slice_value(slice: &Expr) -> impl Iterator<Item = &Expr> {
     match slice {
-        Expr::Tuple(ast::ExprTuple { elts: elements, .. }) => elements.iter().collect(),
-        _ => vec![slice],
+        Expr::Tuple(ast::ExprTuple { elts: elements, .. }) => Left(elements.iter()),
+        _ => Right(std::iter::once(slice)),
     }
 }
 
@@ -77,10 +78,14 @@ enum TypingTarget<'a> {
 
     /// A `typing.Union` type or `|` separated types e.g., `Union[int, str]`
     /// or `int | str`.
-    Union(Vec<&'a Expr>),
+    Union(&'a Expr),
+
+    /// A `typing.Union` type or `|` separated types e.g., `Union[int, str]`
+    /// or `int | str`.
+    Pep604Union(&'a Expr),
 
     /// A `typing.Literal` type e.g., `Literal[1, 2, 3]`.
-    Literal(Vec<&'a Expr>),
+    Literal(&'a Expr),
 
     /// A `typing.Optional` type e.g., `Optional[int]`.
     Optional(&'a Expr),
@@ -110,13 +115,13 @@ impl<'a> TypingTarget<'a> {
                 if semantic.match_typing_expr(value, "Optional") {
                     Some(TypingTarget::Optional(slice.as_ref()))
                 } else if semantic.match_typing_expr(value, "Literal") {
-                    Some(TypingTarget::Literal(resolve_slice_value(slice)))
+                    Some(TypingTarget::Literal(slice))
                 } else if semantic.match_typing_expr(value, "Union") {
-                    Some(TypingTarget::Union(resolve_slice_value(slice)))
+                    Some(TypingTarget::Union(slice))
                 } else if semantic.match_typing_expr(value, "Annotated") {
                     resolve_slice_value(slice.as_ref())
-                        .first()
-                        .map(|&expr| TypingTarget::Annotated(expr))
+                        .next()
+                        .map(|expr| TypingTarget::Annotated(expr))
                 } else {
                     semantic.resolve_call_path(value).map_or(
                         // If we can't resolve the call path, it must be defined
@@ -132,9 +137,7 @@ impl<'a> TypingTarget<'a> {
                     )
                 }
             }
-            Expr::BinOp(..) => Some(TypingTarget::Union(
-                PEP604UnionIterator::new(expr).collect(),
-            )),
+            Expr::BinOp(..) => Some(TypingTarget::Pep604Union(expr)),
             Expr::Constant(ast::ExprConstant {
                 value: Constant::None,
                 ..
@@ -179,7 +182,7 @@ impl<'a> TypingTarget<'a> {
             | TypingTarget::Object
             | TypingTarget::Unknown => true,
             TypingTarget::Known => false,
-            TypingTarget::Literal(elements) => elements.iter().any(|element| {
+            TypingTarget::Literal(element) => resolve_slice_value(element).any(|element| {
                 // Literal can only contain `None`, a literal value, other `Literal`
                 // or an enum value.
                 match TypingTarget::try_from_expr(element, semantic, locator, minor_version) {
@@ -190,12 +193,20 @@ impl<'a> TypingTarget<'a> {
                     _ => false,
                 }
             }),
-            TypingTarget::Union(elements) => elements.iter().any(|element| {
+            TypingTarget::Union(element) => resolve_slice_value(element).any(|element| {
                 TypingTarget::try_from_expr(element, semantic, locator, minor_version)
                     .map_or(true, |new_target| {
                         new_target.contains_none(semantic, locator, minor_version)
                     })
             }),
+            TypingTarget::Pep604Union(element) => {
+                PEP604UnionIterator::new(element).any(|element| {
+                    TypingTarget::try_from_expr(element, semantic, locator, minor_version)
+                        .map_or(true, |new_target| {
+                            new_target.contains_none(semantic, locator, minor_version)
+                        })
+                })
+            }
             TypingTarget::Annotated(element) => {
                 TypingTarget::try_from_expr(element, semantic, locator, minor_version)
                     .map_or(true, |new_target| {
@@ -226,12 +237,20 @@ impl<'a> TypingTarget<'a> {
             | TypingTarget::Object
             | TypingTarget::Known
             | TypingTarget::Unknown => false,
-            TypingTarget::Union(elements) => elements.iter().any(|element| {
+            TypingTarget::Union(element) => resolve_slice_value(element).any(|element| {
                 TypingTarget::try_from_expr(element, semantic, locator, minor_version)
                     .map_or(true, |new_target| {
                         new_target.contains_any(semantic, locator, minor_version)
                     })
             }),
+            TypingTarget::Pep604Union(element) => {
+                PEP604UnionIterator::new(element).any(|element| {
+                    TypingTarget::try_from_expr(element, semantic, locator, minor_version)
+                        .map_or(true, |new_target| {
+                            new_target.contains_any(semantic, locator, minor_version)
+                        })
+                })
+            }
             TypingTarget::Annotated(element) | TypingTarget::Optional(element) => {
                 TypingTarget::try_from_expr(element, semantic, locator, minor_version)
                     .map_or(true, |new_target| {
```

Reply:
This lets you avoid any allocations 😍

Reply:
Oh wow! That's neat :)

We could even go further and merge `Union` and `|` separated unions together in the slice resolver but then that would need to an enum as `itertools::Either` is only for 2 sum types. It would be kinda similar to this implementation except that the duplication of `PEP604UnionIterator` call will be avoided at the expense of having an additional type.

I think this separation is ok for now.